src/layers/core/grid-layer/grid-layer.js

```diff
@@ -54,25 +56,62 @@ function _needsReProjectPoints(oldProps, props) {
   return oldProps.cellSize !== props.cellSize;
 }
 
+function _percentileChanged(oldProps, props) {
+  return oldProps.lowerPercentile !== props.lowerPercentile ||
+    oldProps.upperPercentile !== props.upperPercentile;
+}
+
+function _needsReSortBins(oldProps, props) {
+  return oldProps.getColorValue !== props.getColorValue;
+}
+
 export default class GridLayer extends CompositeLayer {
   initializeState() {
     this.state = {
       layerData: [],
-      countRange: null
+      sortedBins: null,
+      valueDomain: null
     };
   }
 
   updateState({oldProps, props, changeFlags}) {
     if (changeFlags.dataChanged || _needsReProjectPoints(oldProps, props)) {
-      const {data, cellSize, getPosition} = this.props;
+      // project data into hexagons, and get sortedBins
+      this.getLayerData();
+      this.getSortedBins();
+
+      // this needs sortedBins to be set
+      this.onPercentileChange();
+    } else if (_needsReSortBins(oldProps, props)) {
+
+      this.getSortedBins();
+      this.onPercentileChange();
 
-      const {layerData, countRange} =
-        pointToDensityGridData(data, cellSize, getPosition);
+    } else if (_percentileChanged(oldProps, props)) {
 
-      Object.assign(this.state, {layerData, countRange});
+      this.onPercentileChange();
```

Comment:
`onPercentileChanged` is quite suggestive of a callback (i.e. calling a function from another component). I'd change this to `percenticlesChanged();` or similar